### PR TITLE
Add ctors that serialize

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/issues/14118

### CODE REVIEW

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs [hunk 1]
```diff
@@ -73,6 +74,34 @@ public BinaryData(string data)
...
         }
 
+        /// <summary>
+        /// Creates a binary data instance from an object and serializes it
+        /// using the <see cref="JsonObjectSerializer"/>.
+        /// </summary>
+        /// <param name="data">The data that will be serialized.</param>
+        /// <param name="type">The type of the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public BinaryData(object data, Type type) :
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:84]: This doesn't at all look like something that would do JsonSerialization from the outside.

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs [hunk 2]
```diff
@@ -73,6 +74,34 @@ public BinaryData(string data)
...
+        /// <summary>
+        /// Creates a binary data instance from an object and serializes it
+        /// using the provided <see cref="ObjectSerializer"/>.
+        /// </summary>
+        /// <param name="data">The data that will be serialized.</param>
+        /// <param name="type">The type of the data.</param>
+        /// <param name="serializer">The serializer to serialize
+        /// the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public BinaryData(object data, Type type, ObjectSerializer serializer)
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:98]: What about 
>This should not expose ObjectSerializer parameter as that would mean we would need to support async, which we can't do in the constructor. Using this constructor would use JsonObjectSerializer.
?

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs (hunk 1 above)
Comment (pakrym) [reply sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:84]: This is harder to use than the generic overload and clutters a valuable space of public ctors. Why do we think it's worth it?

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:84]: The main reason would be discoverability, but I agree that it is somewhat of a high price to pay in that we now have two ways of doing the same thing.
/cc @KrzysztofCwalina

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:98]: I think this is a reasonable compromise - using ObjectSerializer requires the static methods.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:98]: Default to null - use JsonObjectSerializer.
Remove other ctor.

**Comment by JoshLove-msft** [comment]
@pakrym pinging on this

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs [hunk 3]
```diff
@@ -73,6 +74,24 @@ public BinaryData(string data)
...
+        /// </summary>
+        /// <param name="data">The data that will be serialized.</param>
+        /// <param name="type">The type of the data.</param>
+        /// <param name="serializer">The serializer to serialize
+        /// the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public BinaryData(object data, Type type, ObjectSerializer? serializer = default)
+        {
+            using var memoryStream = new MemoryStream();
+            serializer ??= new JsonObjectSerializer();
```
Comment (pakrym) [inline sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:90]: File a bug to make this less allocaling

**Review by pakrym — APPROVED** [review]

File: sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs:90]: https://github.com/Azure/azure-sdk-for-net/issues/14346

**Pull request merged**